File: app/src/main/java/com/kota/Bahamut/Pages/Model/MailBoxPageHandler.java
package com.kota.Bahamut.Pages.Model;

import com.kota.Telnet.Model.TelnetRow;
import com.kota.Telnet.TelnetClient;
import com.kota.Telnet.TelnetUtils;

public class MailBoxPageHandler {
    private static MailBoxPageHandler _instance = null;

    private MailBoxPageHandler() {
    }

    public static MailBoxPageHandler getInstance() {
        if (_instance == null) {
            _instance = new MailBoxPageHandler();
        }
        return _instance;
    }

    public MailBoxPageBlock load() {
        MailBoxPageBlock board_data = MailBoxPageBlock.create();
        int end_index = 3 + 20;
        for (int i = 3; i < end_index; i++) {
            TelnetRow row = TelnetClient.getModel().getRow(i);
            String article_selected = row.getSpaceString(0, 0).trim();
            int article_index = TelnetUtils.getIntegerFromData(row, 1, 5);
            if (article_index != 0) {
                if (article_selected.length() > 0 && article_selected.charAt(0) == '>') {
                    board_data.selectedItemNumber = article_index;
                }
                byte info = row.data[6];
                String date = row.getSpaceString(8, 12).trim();
                String author = row.getSpaceString(14, 25).trim();
                String origin_mark = row.getSpaceString(27, 28).trim();
                String title = row.getSpaceString(30, 79).trim();
                MailBoxPageItem item = MailBoxPageItem.create();
                if (i == 3) {
                    board_data.minimumItemNumber = article_index;
                }
                board_data.maximumItemNumber = article_index;
                item.Number = article_index;
                item.Date = date;
                item.Author = author;
                item.isRead = (info == 43 || info == 77) ? false : true;
                item.isReply = info == 114 || info == 82;
                item.isMarked = info == 109 || info == 82 || info == 77;
                item.Title = title;
                item.isOrigin = origin_mark.equals("◇") || origin_mark.equals("◆");
                board_data.setItem(i - 3, item);
            }
        }
        return board_data;
    }
}

File: app/src/main/java/com/kota/Bahamut/pages/model/MailBoxPageHandler.kt
package com.kota.Bahamut.pages.model

import com.kota.telnet.model.TelnetRow
import com.kota.telnet.TelnetClient
import com.kota.telnet.TelnetUtils

class MailBoxPageHandler private constructor() {
    fun load(): MailBoxPageBlock {
        val boardData = MailBoxPageBlock.create()
        val endIndex = 3 + 20
        for (i in 3..<endIndex) {
            val row: TelnetRow = TelnetClient.model.getRow(i)!!
            val articleSelected = row.getSpaceString(0, 0).trim { it <= ' ' }
            val articleIndex = TelnetUtils.getIntegerFromData(row, 1, 5)
            if (articleIndex != 0) {
                if (articleSelected.isNotEmpty() && articleSelected[0] == '>') {
                    boardData.selectedItemNumber = articleIndex
                }
                val info = row.data[6]
                val date = row.getSpaceString(8, 12).trim { it <= ' ' }
                val author = row.getSpaceString(14, 25).trim { it <= ' ' }
                val originMark = row.getSpaceString(27, 28).trim { it <= ' ' }
                val title = row.getSpaceString(30, 79).trim { it <= ' ' }
                val item: MailBoxPageItem = MailBoxPageItem.Companion.create()
                if (i == 3) {
                    boardData.minimumItemNumber = articleIndex
                }
                boardData.maximumItemNumber = articleIndex
                item.itemNumber = articleIndex
                item.date = date
                item.author = author
                item.isRead = info.toInt() != 43 && info.toInt() != 77
                item.isReply = info.toInt() == 114 || info.toInt() == 82
                item.isMarked = info.toInt() == 109 || info.toInt() == 82 || info.toInt() == 77
                item.title = title
                item.isOrigin = originMark == "◇" || originMark == "◆"
                boardData.setItem(i - 3, item)
            }
        }
        return boardData
    }

    companion object {
        private var _instance: MailBoxPageHandler? = null

        @JvmStatic
        val instance: MailBoxPageHandler
            get() {
                if (_instance == null) {
                    _instance = MailBoxPageHandler()
                }
                return _instance!!
            }
    }
}
